src/hotspot/share/runtime/arguments.cpp

```diff
@@ -3564,9 +3564,9 @@ void Arguments::init_shared_archive_paths() {
           // If +AutoCreateSharedArchive and the specified shared archive does not exist,
           // regenerate the dynamic archive base on default archive.
           if (AutoCreateSharedArchive && !os::file_exists(SharedArchiveFile)) {
-            SharedDynamicArchivePath = const_cast<char *>(SharedArchiveFile);
-            SharedArchivePath = get_default_shared_archive_path();
             DynamicDumpSharedSpaces = true;
+            ArchiveClassesAtExit = const_cast<char *>(SharedArchiveFile);
+            SharedArchivePath = get_default_shared_archive_path();
```

Comment:
I think we should set SharedArchiveFile=NULL here. That way, when foo.jsa doesn't exist, the two command lines are equivalent:

- java -XX:ArchiveClassesAtExit=foo.jsa ...
- java -XX:SharedArchiveFile=foo.jsa -XX:+AutoCreateSharedArchive ...

and we can revert the changes on line 3602, to make the code easier to understand.